### PR TITLE
Simultaneously support versions v2 and v2beta2 of Autoscaling

### LINKED ISSUE #943: Update HorizontalPodAutoscaler code from v1
The current HPA code uses version v1 of the API.  We should upgrade to v2 or v2beta2 depending on what is supported in the versions of kubernetes that we need to support.

### PR DESCRIPTION
**Pull request by kevinearls**
Signed-off-by: Kevin Earls <kearls@redhat.com>

This is to resolve #943.  Choose which version of autoscaling to use depending on what is available at runtime.  We need to be able to support both v2 and v2beta2

### CODE REVIEW

**Comment by kevinearls** [comment]
@pavolloffay Please review.  BTW should I need to ask for a review if it already says "open-telemetry/operator-approvers was requested for review".  I don't want to be spamming you unnecessarily.

File: pkg/collector/reconcile/horizontalpodautoscaler.go [hunk 1]
```diff
@@ -51,66 +54,124 @@ func HorizontalPodAutoscalers(ctx context.Context, params Params) error {
 	return nil
 }
 
-func expectedHorizontalPodAutoscalers(ctx context.Context, params Params, expected []autoscalingv1.HorizontalPodAutoscaler) error {
+func expectedHorizontalPodAutoscalers(ctx context.Context, params Params, expected []runtime.Object) error {
```
[the remark this answers is not in the document]
Comment (pavolloffay) [reply pkg/collector/reconcile/horizontalpodautoscaler.go:57]: Something like the following could work: 

```golang
func expectedHorizontalPodAutoscalers(ctx context.Context, params Params, expected []runtime.Object) error {
	autoscalingVersion := params.Config.AutoscalingVersion()
	var existing client.Object
	if autoscalingVersion == config.AutoscalingVersionV2Beta2 {
		existing = &autoscalingv2beta2.HorizontalPodAutoscaler{}
	} else {
		existing = &autoscalingv2.HorizontalPodAutoscaler{}
	}

	for _, obj := range expected {
		desired, _ := meta.Accessor(obj)

		if err := controllerutil.SetControllerReference(&params.Instance, desired, params.Scheme); err != nil {
			return fmt.Errorf("failed to set controller reference: %w", err)
		}

		nns := types.NamespacedName{Namespace: desired.GetNamespace(), Name: desired.GetName()}
		err := params.Client.Get(ctx, nns, existing)
		if k8serrors.IsNotFound(err) {
			if err := params.Client.Create(ctx, obj.(client.Object)); err != nil {
				return fmt.Errorf("failed to create: %w", err)
			}
			params.Log.V(2).Info("created", "hpa.name", desired.GetName(), "hpa.namespace", desired.GetNamespace())
			continue
		} else if err != nil {
			return fmt.Errorf("failed to get %w", err)
		}

		existing.SetOwnerReferences(desired.GetOwnerReferences())
		setAutoscalerSpec(params.Instance, existing)

		annos := existing.GetAnnotations()
		for k, v := range desired.GetAnnotations() {
			annos[k] = v
		}
		existing.SetAnnotations(annos)
		labels := existing.GetLabels()
		for k, v := range desired.GetLabels() {
			labels[k] = v
		}
		existing.SetLabels(labels)

		patch := client.MergeFrom(existing)

		if err := params.Client.Patch(ctx, existing, patch); err != nil {
			return fmt.Errorf("failed to apply changes: %w", err)
		}

		params.Log.V(2).Info("applied", "hpa.name", desired.Name, "hpa.namespace", desired.Namespace)
	}

	return nil
}
```

File: pkg/autodetect/main.go [hunk 2]
```diff
@@ -64,3 +68,29 @@ func (a *autoDetect) Platform() (platform.Platform, error) {
...
+			sort.Slice(versions, func(i, j int) bool {
+				return versions[i].Version < versions[j].Version
+			})
+
+			for _, version := range versions {
+				if version.Version == "v2" || version.Version == "v2beta2" { // Can't use the constants from internal/config/main.go as that would create an import cycle
+					return version.Version, nil
+				}
+			}
+			return "", errors.New("Failed to find appropriate version of apiGroup autoscaling")
```
Comment (pavolloffay) [inline pkg/autodetect/main.go:91]: We could include which versions are "appropriate".

File: pkg/collector/horizontalpodautoscaler.go [hunk 3]
```diff
@@ -26,30 +29,78 @@ import (
 
 const defaultCPUTarget int32 = 90
 
-func HorizontalPodAutoscaler(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector) autoscalingv1.HorizontalPodAutoscaler {
+func HorizontalPodAutoscaler(cfg config.Config, logger logr.Logger, otelcol v1alpha1.OpenTelemetryCollector) runtime.Object {
```
Comment (pavolloffay) [inline pkg/collector/horizontalpodautoscaler.go:32]: instread of returning `runtime.Object` wouldn't it be better to return `client.Object`? That way we don't have to cast this in the reconcile.

File: internal/config/main.go [hunk 4]
```diff
@@ -132,6 +138,14 @@ func (c *Config) AutoDetect() error {
 		}
 	}
 
+	hpaVersion, err := c.autoDetect.HPAVersion()
+	if err != nil {
+		return err
+	} else {
```
Comment (pavolloffay) [inline internal/config/main.go:144]: no need for else branch

File: pkg/collector/reconcile/horizontalpodautoscaler.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply pkg/collector/reconcile/horizontalpodautoscaler.go:57]: @kevinearls did you try this de-duplication?

**Comment by kevinearls** [comment]
@pavolloffay Yes, I have now added the de duplication code.  The original code you posted required some changes.

File: internal/config/main.go [hunk 5]
```diff
@@ -30,6 +30,9 @@ const (
 	defaultAutoDetectFrequency           = 5 * time.Second
 	defaultCollectorConfigMapEntry       = "collector.yaml"
 	defaultTargetAllocatorConfigMapEntry = "targetallocator.yaml"
+	AutoscalingVersionV2                 = "v2"
```
Comment (pavolloffay) [inline internal/config/main.go:33]: nit: define these as enum (a golang type)

Comment (kevinearls) [reply internal/config/main.go:33]: @pavolloffay Can you point me to an example of what you want here?    Googling "golang enums" returns a bunch of disparate answers

**Comment by kevinearls** [comment]
@pavolloffay Can you review this?  I'd like to get this merged before you go on PTO.

File: controllers/opentelemetrycollector_controller.go [hunk 6]
```diff
@@ -173,14 +175,25 @@ func (r *OpenTelemetryCollectorReconciler) RunTasks(ctx context.Context, params
 
 // SetupWithManager tells the manager what our controller is interested in.
 func (r *OpenTelemetryCollectorReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
+	err := r.config.AutoDetect() // We need to call this so we can get the correct autodetect version
+	if err != nil {
+		return err
+	}
+	autoscalingVersion := r.config.AutoscalingVersion()
```
Comment (pavolloffay) [inline controllers/opentelemetrycollector_controller.go:182]: nit: move this closer to the place where it is used - line 192

File: internal/config/main_test.go [hunk 7]
```diff
@@ -102,6 +102,10 @@ type mockAutoDetect struct {
 	PlatformFunc func() (platform.Platform, error)
 }
 
+func (m *mockAutoDetect) HPAVersion() (autodetect.AutoscalingVersion, error) {
+	return autodetect.DefaultAutoscalingVersion, nil // TODO add a test?
```
Comment (pavolloffay) [inline internal/config/main_test.go:106]: nit: remove todo?

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**